content/blog/2019-03-24-tdd-immutable-tests/index.md

```diff
@@ -0,0 +1,188 @@
+---
+title: 'TDD: Your Unit Tests Should Be Immutable'
+date: '2019-03-24T12:51:00.314Z'
+---
+
+# Picture this...
+
+- Your project has been following TDD from the very start
+- You're tasked with refactoring a function (to improve performance, say)
+- You make a relatively minor change, and run the tests
+- Your tests for that unit pass, but then a bunch of unit tests in some random component fail!
+  - "What the hell is happening, these components aren't even related!" I hear you say.
+
+To fix it you will need to change those tests in those unrelated units, and that's a big problem. Your tests need to be _immutable to refactoring_ and the above scenario is a `Test Smell`.
+
+# Why is this a problem?
+
+There are lots of benefits to TDD, but one that doesn't get as much airtime is that:
+
+> A good test-suite gives you **confidence** about the behaviour of your system.
+
+However that confidence is NOT guaranteed by tooling, it comes from the fact that you followed TDD. You wrote a test, it failed, then you wrote some code and it passed. Therefore, that code works.
+
+That means that the only reason you trust your codebase is because you trust your fellow developers to have _also_ followed TDD. Which is to say, that relying on your test-suite is not a matter of Fact, but a matter of Trust.
+
+Hence once a test-suite exists, we need to protect our Trust in it by minimizing changes to it.
+
+Coming back to our hypothetical situation from earlier, Refactoring (by definition) is not supposed to be a change in behaviour. If you're now forced to update test-code in unrelated units, you now need to _manually_ validate those units and (all their usages) before trust is restored in the system.
+
+# Reasons for Problem
+
+For example, your unit originally set an internal state's `canProceed` and you refactored it to `shouldProceed`. These other units are failing because they were asserting on the value of `canProceed`.
+
+If you go into those failing tests and update the tests with the new assumption it's _highly_ likely that an error will get made and you might assert `false` instead of `true` or any number of other issues. You might even accidentally delete a test (in more complex scenarios). The worst part of it is that your test suite can't save you from these mistakes because it can't self-test; the Light going Green doesn't mean anything if the test implementation is giving you a false-positive.
+
+# Desired Behaviour
+
+The only reason to change a test should be if the required behaviour of the system changes. In all other cases, we expect the tests to be `Immutable`.
```

Comment:
From tushar: 

Also there are two case you might want to change your tests —

1. Its a breaking change,
2. There is a bug.